### ISSUE TITLE
Instance ShaderMaterials going black after 4.1k instances

### ISSUE THREAD

**Issue by BurningFluffer** [godotengine/godot#79167]
### Godot version

4.1.stable

### System information

Godot v4.1.stable - Windows 10.0.19044 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3050 (NVIDIA; 31.0.15.2824) - Intel(R) Core(TM) i5-4690K CPU @ 3.50GHz (4 Threads)

### Issue description

General: I can spawn blocks with 6 meshes, avg 2-3 materials each. Those are _instance spacial shader materials_, just hue parameters mixed with a texture (different texture -> different materials, but hues and wether texture is used varies between instances). 

Problem: once I spawn ~680-700 blocks (4.1k meshes), all following blocks spawn with pure black texture. it doesn't matter if new parameters or old ones are used. I can despawn a few random colored blocks and that's exactly how many more colored meshes I will be able to spawn. Other than that, the object is lit by environmental light like all other.
![image](https://github.com/godotengine/godot/assets/138900251/dbcffd68-c2f3-41ee-9ca6-1f97daa04129)

That said, draw calls are under 100, fps stays at 60 and I can spawn 5 times more blocks before any loss of fps or general performance. Somehow, all materials on a mesh turn black and ignore parameters.

### Steps to reproduce

1. Load the project and an auto-clicker.
2. Star the game, wail with auto clicker on 10 ms or so.
3. (Extra: Some block textures flicker, but return to normal)
4. Around 4100 blocks (4,1k meshes, same as when it fails in my main project) you will see that all new meshes are black.

### Minimal reproduction project

[Bug-ShaderMaterialInstanceLimit.zip](https://github.com/godotengine/godot/files/11984381/Bug-ShaderMaterialInstanceLimit.zip)

**Comment by clayjohn** [comment]
Do you get any errors printed in the console when the issue occurs?

**Comment by BurningFluffer** [comment]
![image](https://github.com/godotengine/godot/assets/138900251/51931a73-d3eb-43f5-974a-eb811398cc19)
It says too many istances are using shader instance variables, 4096 being the limit, which is 2^12. Though, there are onl a few options for colors, and int he main progect, I can just use the same parameters for all meshes and still hit this limit. Is it creating a new material with each mesh, regardless of weather it needs to? 
Also there is a problem of random meshes changing their color though they are not supposed to be related.

**Comment by clayjohn** [comment]
Have you tried increasing the limit as the error suggests doing?

**Comment by BurningFluffer** [comment]
Yeah, It allows as many times more as bigger the buffer is set. However, by that time, the world becomes a flickering sea of candy as each new block changes the colors of many other blocks.

**Comment by bitsawer** [comment]
Seems like the original issue is solved and not really an issue, just increase the buffer size in the project settings. However, I noticed two other issues that might be bugs or something else. I uploaded a simplified version of the MRP project that has some useless stuff commented out and doesn't require an auto-clicker to actually see the problem:

[Bug-ShaderMaterialInstanceLimit-simpler.zip](https://github.com/godotengine/godot/files/12016135/Bug-ShaderMaterialInstanceLimit-simpler.zip)

To test the project, run it and click the mouse so that the raycast hits the ground plane, hold the button down and move the mouse. You can "paint" new meshes on the plane that have all instance unfiform colors. You can see two potential issues:

1) The material of a new box mesh is black during the first frame when it is created. After that, the instance uniforms are updated and the box gets its bright color. Not sure if this is intentional, but it can be pretty bad if you use instance uniforms to index into an uniform array or texture etc. During this first frame you can accidentally read garbage if you don't guard all your reads, which can affect performance and can be tricky to do. Brute forcing a call to `MaterialStorage::_update_global_shader_uniforms()`  in `MaterialStorage::global_shader_parameters_instance_update()` with some check removes seems to work and gets rid of this possible glitch, but there should be a better and more performant way to do this.

2) After a few hundred boxes are added, it looks like some of the old instance colors are overriden or corrupted. As far as I can see it, the demo only modified instance colors of the newly created box meshes, but occasionally you can see how some of the old box colors change, too. The more boxes you paint, the more easier this is to see as many old mesh colors start to glitch out.

**Comment by bitsawer** [comment]
Turns out the solution to issue 2. I mentioned is to simply enable "Local to Scene" for the `Block` MeshInstance3D in the `block.tscn` so that they don't conflict with each other. You can read more about this flag and PackedScenes in the documentation. Or you can manually duplicate resources or nodes in code during runtime, but that can result in more heavy-handed results.

My comment about the delayed instance uniform update resulting in one-frame black material is still valid, but I'll do a separate, new bug report if I can't find any existing bug reports. Let us know if something other than that issue is still unsolved here, otherwise I'll close this issue as resolved at some point in the near future.

In short, enable this checkbox:

![local](https://github.com/godotengine/godot/assets/22456603/fc7893e5-ce39-46ae-a7c5-920375a645a9)

**Comment by BurningFluffer** [comment]
Thank you! I currently see no other related issues. Have a great day!

**Comment by bitsawer** [comment]
Closing as promised, the new found issue is now tracked in https://github.com/godotengine/godot/issues/79602
